File: packages/vue-im-ui/src/shared/contracts/message-content-contract.ts
export type MessageContentKind =
  | "text" | "richText" | "markdown" | "code" | "image" | "multiImage"
  | "video" | "audio" | "file" | "location" | "contactCard" | "linkPreview"
  | "poll" | "task" | "calendarEvent" | "miniApp" | "topic" | "system"
  | "notice" | "forward" | "mergedForward" | "reply" | "threadRoot"
  | "ephemeral" | "readOnce" | "burnAfterRead";

export type MessageContentFamily =
  | "text" | "richText" | "code" | "image" | "video" | "audio" | "file"
  | "location" | "card" | "link" | "poll" | "task" | "calendar" | "miniApp"
  | "topic" | "system" | "forward" | "reply" | "thread" | "treatment";

export type MessageContentCapability =
  | "select" | "copy" | "openLink" | "horizontalScroll" | "open" | "save"
  | "retry" | "zoom" | "swipe" | "play" | "pause" | "fullscreen" | "seek"
  | "reveal" | "vote" | "toggleTask" | "jumpToOriginal" | "openThread" | "openOnce";

export interface MessageContentContract {
  kind: MessageContentKind;
  wireType?: string;
  family: MessageContentFamily;
  capabilities: ReadonlySet<MessageContentCapability>;
}

const definitions: Record<MessageContentKind, readonly [string | undefined, MessageContentFamily, readonly MessageContentCapability[]]> = {
  text: ["text", "text", ["select", "copy"]],
  richText: ["rich_text", "richText", ["select", "copy", "openLink"]],
  markdown: ["rich_text", "richText", ["select", "copy", "openLink"]],
  code: ["rich_text", "code", ["select", "copy", "horizontalScroll"]],
  image: ["image", "image", ["open", "save", "retry", "zoom"]],
  multiImage: ["image_group", "image", ["open", "save", "retry", "zoom", "swipe"]],
  video: ["video", "video", ["play", "pause", "fullscreen", "save", "retry"]],
  audio: ["audio", "audio", ["play", "pause", "seek", "retry"]],
  file: ["file", "file", ["open", "save", "reveal", "retry"]],
  location: ["location", "location", ["open"]],
  contactCard: ["card", "card", ["open"]],
  linkPreview: ["link_card", "link", ["open", "copy"]],
  poll: ["vote", "poll", ["vote"]],
  task: ["task", "task", ["toggleTask", "open"]],
  calendarEvent: ["schedule", "calendar", ["open"]],
  miniApp: ["mini_program", "miniApp", ["open"]],
  topic: ["custom", "topic", ["open"]],
  system: ["system", "system", []],
  notice: ["notification", "system", ["open"]],
  forward: ["forward", "forward", ["open"]],
  mergedForward: ["forward", "forward", ["open"]],
  reply: ["quote", "reply", ["jumpToOriginal"]],
  threadRoot: ["thread", "thread", ["openThread"]],
  ephemeral: [undefined, "treatment", []],
  readOnce: [undefined, "treatment", ["openOnce"]],
  burnAfterRead: [undefined, "treatment", ["openOnce"]],
};

export const MESSAGE_CONTENT_KINDS = Object.freeze(Object.keys(definitions) as MessageContentKind[]);

export function resolveMessageContentContract(kind: MessageContentKind): MessageContentContract {
  const [wireType, family, capabilities] = definitions[kind];
  return { kind, wireType, family, capabilities: new Set(capabilities) };
}

const wireDefaults: Record<string, MessageContentKind> = {
  text: "text", rich_text: "richText", image: "image", image_group: "multiImage",
  video: "video", audio: "audio", file: "file", location: "location", card: "contactCard",
  link_card: "linkPreview", vote: "poll", task: "task", schedule: "calendarEvent",
  mini_program: "miniApp", system: "system", notification: "notice", forward: "forward",
  quote: "reply", thread: "threadRoot", custom: "topic",
};

/** Maps host content kinds to default renderers. Hosts may refine rich/custom subtypes. */
export function messageContentKindFromWire(wireType: string): MessageContentKind | undefined {
  return wireDefaults[wireType];
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/MessageContentContract.kt
package com.flare.im.ui

enum class FlareMessageContentCapability {
    Select, Copy, OpenLink, HorizontalScroll, Open, Save, Retry, Zoom, Swipe,
    Play, Pause, Fullscreen, Seek, Reveal, Vote, ToggleTask, JumpToOriginal,
    OpenThread, OpenOnce,
}

enum class FlareMessageContentFamily {
    Text, RichText, Code, Image, Video, Audio, File, Location, Card, Link,
    Poll, Task, Calendar, MiniApp, Topic, System, Forward, Reply, Thread, Treatment,
}

enum class FlareMessageContentKind(
    val wireType: String?,
    val family: FlareMessageContentFamily,
    val capabilities: Set<FlareMessageContentCapability>,
) {
    Text("text", FlareMessageContentFamily.Text, setOf(FlareMessageContentCapability.Select, FlareMessageContentCapability.Copy)),
    RichText("rich_text", FlareMessageContentFamily.RichText, setOf(FlareMessageContentCapability.Select, FlareMessageContentCapability.Copy, FlareMessageContentCapability.OpenLink)),
    Markdown("rich_text", FlareMessageContentFamily.RichText, setOf(FlareMessageContentCapability.Select, FlareMessageContentCapability.Copy, FlareMessageContentCapability.OpenLink)),
    Code("rich_text", FlareMessageContentFamily.Code, setOf(FlareMessageContentCapability.Select, FlareMessageContentCapability.Copy, FlareMessageContentCapability.HorizontalScroll)),
    Image("image", FlareMessageContentFamily.Image, setOf(FlareMessageContentCapability.Open, FlareMessageContentCapability.Save, FlareMessageContentCapability.Retry, FlareMessageContentCapability.Zoom)),
    MultiImage("image_group", FlareMessageContentFamily.Image, setOf(FlareMessageContentCapability.Open, FlareMessageContentCapability.Save, FlareMessageContentCapability.Retry, FlareMessageContentCapability.Zoom, FlareMessageContentCapability.Swipe)),
    Video("video", FlareMessageContentFamily.Video, setOf(FlareMessageContentCapability.Play, FlareMessageContentCapability.Pause, FlareMessageContentCapability.Fullscreen, FlareMessageContentCapability.Save, FlareMessageContentCapability.Retry)),
    Audio("audio", FlareMessageContentFamily.Audio, setOf(FlareMessageContentCapability.Play, FlareMessageContentCapability.Pause, FlareMessageContentCapability.Seek, FlareMessageContentCapability.Retry)),
    File("file", FlareMessageContentFamily.File, setOf(FlareMessageContentCapability.Open, FlareMessageContentCapability.Save, FlareMessageContentCapability.Reveal, FlareMessageContentCapability.Retry)),
    Location("location", FlareMessageContentFamily.Location, setOf(FlareMessageContentCapability.Open)),
    ContactCard("card", FlareMessageContentFamily.Card, setOf(FlareMessageContentCapability.Open)),
    LinkPreview("link_card", FlareMessageContentFamily.Link, setOf(FlareMessageContentCapability.Open, FlareMessageContentCapability.Copy)),
    Poll("vote", FlareMessageContentFamily.Poll, setOf(FlareMessageContentCapability.Vote)),
    Task("task", FlareMessageContentFamily.Task, setOf(FlareMessageContentCapability.ToggleTask, FlareMessageContentCapability.Open)),
    CalendarEvent("schedule", FlareMessageContentFamily.Calendar, setOf(FlareMessageContentCapability.Open)),
    MiniApp("mini_program", FlareMessageContentFamily.MiniApp, setOf(FlareMessageContentCapability.Open)),
    Topic("custom", FlareMessageContentFamily.Topic, setOf(FlareMessageContentCapability.Open)),
    System("system", FlareMessageContentFamily.System, emptySet()),
    Notice("notification", FlareMessageContentFamily.System, setOf(FlareMessageContentCapability.Open)),
    Forward("forward", FlareMessageContentFamily.Forward, setOf(FlareMessageContentCapability.Open)),
    MergedForward("forward", FlareMessageContentFamily.Forward, setOf(FlareMessageContentCapability.Open)),
    Reply("quote", FlareMessageContentFamily.Reply, setOf(FlareMessageContentCapability.JumpToOriginal)),
    ThreadRoot("thread", FlareMessageContentFamily.Thread, setOf(FlareMessageContentCapability.OpenThread)),
    Ephemeral(null, FlareMessageContentFamily.Treatment, emptySet()),
    ReadOnce(null, FlareMessageContentFamily.Treatment, setOf(FlareMessageContentCapability.OpenOnce)),
    BurnAfterRead(null, FlareMessageContentFamily.Treatment, setOf(FlareMessageContentCapability.OpenOnce)),
}

data class FlareMessageContentContract(
    val kind: FlareMessageContentKind,
    val wireType: String? = kind.wireType,
    val family: FlareMessageContentFamily = kind.family,
    val capabilities: Set<FlareMessageContentCapability> = kind.capabilities,
)

fun resolveMessageContentContract(kind: FlareMessageContentKind) = FlareMessageContentContract(kind)
